sys/net/application_layer/gcoap/gcoap.c

```diff
@@ -285,15 +285,15 @@ static size_t _handle_req(coap_pkt_t *pdu, uint8_t *buf, size_t len,
 
     if (coap_get_observe(pdu) == COAP_OBS_REGISTER) {
         int empty_slot = _find_obs_memo(&memo, remote, pdu);
+        if ((resource_memo != NULL)
+                && _endpoints_equal(remote, resource_memo->observer)) {
+            /* observer re-registering with new token */
+            memo = resource_memo;
+            observer = resource_memo->observer;
+        }
```

Comment:
What if a memo is found for the remote+token in _find_obs_memo(), but it doesn't match remote+resource_memo? observer still will be null. Must handle when the provided token already is in use for a different resource.